File: ureka-framework-main/tests/operation/test_03_success_when_transfer_device_ownership.py
######################################################
# Test Fixtures
######################################################
import pytest
from tests.conftest import (
    current_setup_log,
    current_teardown_log,
    current_test_given_log,
    current_test_when_and_then_log,
)
from tests.conftest import (
    create_simulated_comm_connection,
    wait_simulated_comm_completed,
)
from tests.conftest import (
    device_manufacturer_server,
    device_manufacturer_server_and_her_device,
    device_owner_agent,
    device_owner_agent_and_her_device,
    device_owner_agent_and_her_session,
    enterprise_provider_server,
    enterprise_provider_server_and_her_session,
    attacker_server,
)
from ureka_framework.resource.storage.simple_storage import SimpleStorage
from typing import Iterator

######################################################
# Import
######################################################
import ureka_framework.model.message_model.u_ticket as u_ticket


class TestSuccessWhenIntializeAgentOrServer:
    @pytest.fixture(scope="function", autouse=True)
    def setup_teardown(self) -> Iterator[None]:
        # RE-GIVEN: Reset the test environment
        current_setup_log()
        SimpleStorage.delete_storage_in_test()

        # GIVEN+WHEN+THEN:
        yield

        # RE-GIVEN: Reset the test environment
        current_teardown_log()
        SimpleStorage.delete_storage_in_test()

    def test_success_when_apply_ownership_u_ticket_on_device(self) -> None:
        current_test_given_log()

        # GIVEN: Initialized DM's CS and DM's IoTD
        (
            self.cloud_server_dm,
            self.iot_device,
        ) = device_manufacturer_server_and_her_device()

        assert (
            self.iot_device.shared_data.this_device.owner_pub_key_str
            == self.cloud_server_dm.shared_data.this_person.person_pub_key_str
        )

        # GIVEN: Initialized DO's UA
        self.user_agent_do = device_owner_agent()

        # WHEN:
        current_test_when_and_then_log()
        # WHEN: Issuer: DM's CS generate & send the ownership_u_ticket to DO's UA
        create_simulated_comm_connection(self.cloud_server_dm, self.user_agent_do)
        target_device_id = self.iot_device.shared_data.this_device.device_pub_key_str
        original_device_order = self.iot_device.shared_data.this_device.ticket_order
        original_agent_order = self.cloud_server_dm.shared_data.device_table[
            target_device_id
        ].ticket_order
        generated_request: dict = {
            "device_id": f"{target_device_id}",
            "holder_id": f"{self.user_agent_do.shared_data.this_person.person_pub_key_str}",
            "u_ticket_type": f"{u_ticket.TYPE_OWNERSHIP_UTICKET}",
        }
        self.cloud_server_dm.flow_issuer_issue_u_ticket.issuer_issue_u_ticket_to_holder(
            device_id=target_device_id, arbitrary_dict=generated_request
        )
        wait_simulated_comm_completed(self.user_agent_do, self.cloud_server_dm)

        # WHEN: Holder: DO's UA forward the ownership_u_ticket
        create_simulated_comm_connection(self.user_agent_do, self.iot_device)
        self.user_agent_do.flow_apply_u_ticket.holder_apply_u_ticket(target_device_id)
        wait_simulated_comm_completed(self.user_agent_do, self.iot_device)

        # THEN: Succeed to transfer ownership (become DO's IoTD)
        assert "SUCCESS" in self.iot_device.shared_data.result_message
        assert "SUCCESS" in self.user_agent_do.shared_data.result_message
        # THEN: Device: Set New Owner
        assert (
            self.iot_device.shared_data.this_device.owner_pub_key_str
            == self.user_agent_do.shared_data.this_person.person_pub_key_str
        )
        # THEN: Device: Update ticket order, DM's CS cannot access DO's IoTD anymore
        assert (
            self.iot_device.shared_data.this_device.ticket_order
            == original_device_order + 1
        )
        assert (
            self.user_agent_do.shared_data.device_table[target_device_id].ticket_order
            == original_agent_order + 1
        )

        # WHEN: Holder: DO's UA return the ownership_r_ticket to DM's CS
        create_simulated_comm_connection(self.cloud_server_dm, self.user_agent_do)
        self.user_agent_do.flow_issuer_issue_u_ticket.holder_send_r_ticket_to_issuer(
            target_device_id
        )
        wait_simulated_comm_completed(self.cloud_server_dm, self.user_agent_do)

        # THEN: Issuer: DM's CS know that DO's UA has become the new owner of DO's IoTD (& ticket order++)
        assert "SUCCESS" in self.cloud_server_dm.shared_data.result_message
        assert (
            self.cloud_server_dm.shared_data.device_table.get(target_device_id) == None
        )

    def test_success_when_reboot_device(self) -> None:
        current_test_given_log()

        # GIVEN: Initialized DO's UA and DO's IoTD
        (
            self.user_agent_do,
            self.iot_device,
        ) = device_owner_agent_and_her_device()

        # WHEN: Reboot DO's IoTD
        current_test_when_and_then_log()
        self.iot_device.reboot_device()

        # THEN: Still be Initialized DO's IoTD
        assert (
            self.iot_device.shared_data.this_device.owner_pub_key_str
            == self.user_agent_do.shared_data.this_person.person_pub_key_str
        )

File: UrekaApp/app/src/androidTest/java/test/operation/Test03SuccessWhenTransferDeviceOwnership.java
package test.operation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static Conftest.currentSetupLog;
import static Conftest.currentTeardownLog;
import static Conftest.currentTestGivenLog;
import static Conftest.currentTestWhenAndThenLog;
import static Conftest.deviceManufacturerServerAndHerDevice;
import static Conftest.deviceOwnerAgent;
import static Conftest.deviceOwnerAgentAndHerDevice;

import java.util.Map;

import ureka.framework.logic.DeviceController;
import ureka.framework.model.message_model.UTicket;
import ureka.framework.resource.Pair;

public class Test03SuccessWhenTransferDeviceOwnership {
    private DeviceController cloudServerDm;
    private DeviceController iotDevice;
    private DeviceController userAgentDO;

    // RE-GIVEN: Reset the test environment
    // @BeforeEach
    public void setup() {
        currentSetupLog();
        // SimpleStorage.deleteStorageInTest();
    }

    // RE-GIVEN: Reset the test environment
    // @AfterEach
    public void teardown() {
        currentTeardownLog();
        // SimpleStorage.deleteStorageInTest();
    }

    // @Test
    public void testSuccessWhenApplyOwnershipUTicketOnDevice() {
        currentTestGivenLog();

        // GIVEN: Initialized DM's CS and DM's IoTD
        Pair devices = deviceManufacturerServerAndHerDevice();
        this.cloudServerDm = (DeviceController) devices.getPairFirst();
        this.iotDevice = (DeviceController) devices.getPairSecond();

        assertEquals(this.iotDevice.getSharedData().getThisDevice().getOwnerPubKeyStr(), this.cloudServerDm.getSharedData().getThisPerson().getPersonPubKeyStr());

        // GIVEN: Initialized DO's UA
        this.userAgentDO = deviceOwnerAgent();

        // WHEN:
        currentTestWhenAndThenLog();
        // WHEN: Issuer: DM's CS generate & send the ownership_u_ticket to DO's UA
        // createSimulatedCommConnection(this.cloudServerDm,this.userAgentDO);
        String targetDeviceId = this.iotDevice.getSharedData().getThisDevice().getDevicePubKeyStr();
        Integer originalDeviceOrder = this.iotDevice.getSharedData().getThisDevice().getTicketOrder();
        Integer originalAgentOrder = this.cloudServerDm.getSharedData().getDeviceTable().get(targetDeviceId).getTicketOrder();

        Map<String, String> generatedRequest = Map.of(
                "deviceId", targetDeviceId,
                "holderId", this.userAgentDO.getSharedData().getThisPerson().getPersonPubKeyStr(),
                "uTicketType", UTicket.TYPE_OWNERSHIP_UTICKET
        );
        this.cloudServerDm.getFlowIssuerIssueUTicket().issuerIssueUTicketToHolder(targetDeviceId,generatedRequest);
        this.userAgentDO.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(this.userAgentDO,this.cloudServerDm);

        // WHEN: Holder: DO's UA forward the ownershipUTicket
        // createSimulatedCommConnection(this.userAgentDO,this.iotDevice);
        this.userAgentDO.getFlowApplyUTicket().holderApplyUTicket(targetDeviceId);
        this.iotDevice.getMsgReceiver()._recvXxxMessage();
        this.userAgentDO.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(this.userAgentDO,this.iotDevice);

        // THEN: Succeed to transfer ownership (become DO's IoTD)
        assertTrue(this.iotDevice.getSharedData().getResultMessage().contains("SUCCESS"));
        assertTrue(this.userAgentDO.getSharedData().getResultMessage().contains("SUCCESS"));

        // THEN: Device: Set New Owner
        assertEquals(this.iotDevice.getSharedData().getThisDevice().getOwnerPubKeyStr(), this.userAgentDO.getSharedData().getThisPerson().getPersonPubKeyStr());
        // THEN: Device: Update ticket order, DM's CS cannot access DO's IoTD anymore
        assertEquals(originalDeviceOrder + 1, this.iotDevice.getSharedData().getThisDevice().getTicketOrder());
        assertEquals(originalAgentOrder + 1, this.userAgentDO.getSharedData().getDeviceTable().get(targetDeviceId).getTicketOrder());

        // WHEN: Holder: DO's UA return the ownershipRTicket to DM's CS
        // createSimulatedCommConnection(this.cloudServerDm,this.userAgentDO);
        this.userAgentDO.getFlowIssuerIssueUTicket().holderSendRTicketToIssuer(targetDeviceId);
        this.cloudServerDm.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(this.cloudServerDm,this.userAgentDO);

        // THEN: Issuer: DM's CS know that DO's UA has become the new owner of DO's IoTD (& ticket order++)
        assertTrue(this.cloudServerDm.getSharedData().getResultMessage().contains("SUCCESS"));
        assertNull(this.cloudServerDm.getSharedData().getDeviceTable().get(targetDeviceId));
    }

    // Skipped
    // @Test
    public void testSuccessWhenRebootDevice() {
        currentTestGivenLog();

        // GIVEN: Initialized DM's CS and DM's IoTD
        Pair devices = deviceOwnerAgentAndHerDevice();
        this.userAgentDO = (DeviceController) devices.getPairFirst();
        this.iotDevice = (DeviceController) devices.getPairSecond();

        // WHEN: Reboot DM's IoTD
        currentTestWhenAndThenLog();
        this.iotDevice.rebootDevice();

        // THEN: Still be Initialized DM's IoTD
        assertEquals(this.iotDevice.getSharedData().getThisDevice().getOwnerPubKeyStr(), this.userAgentDO.getSharedData().getThisPerson().getPersonPubKeyStr());
    }

    public void runAll() {
        setup();
        testSuccessWhenApplyOwnershipUTicketOnDevice();
        teardown();
    }
}

